### PR TITLE
pallet-collective | port to frame v2

### PR DESCRIPTION
**Pull request by shamb0**
# pallet-collective | port to frame v2 

related #7882

:warning: Breaking Change :warning:

From https://crates.parity.io/frame_support/attr.pallet.html#checking-upgrade-guidelines

```
storages now use PalletInfo for module_prefix instead of the one given to decl_storage: 
Thus any use of this pallet in construct_runtime! should be careful to update name in order 
not to break storage or to upgrade storage (moreover for instantiable pallet). 
If pallet is published, make sure to warn about this breaking change.
```

So users of the collective pallet must be careful about the name they used in construct_runtime!. Hence the runtime-migration label, which might not be needed depending on the configuration of the collective pallet.

### CODE REVIEW

File: frame/collective/src/lib.rs [hunk 1]
```diff
@@ -350,34 +239,27 @@ decl_module! {
...
-		/// # </weight>
-		#[weight = (
+		#[pallet::weight(
 			T::WeightInfo::execute(
 				*length_bound, // B
 				T::MaxMembers::get(), // M
-			).saturating_add(proposal.get_dispatch_info().weight), // P
-			DispatchClass::Operational
+			)
 		)]
```
Comment (thiolliere) [inline frame/collective/src/lib.rs:247]: it should be:
```rust
	#[pallet::weight((
		T::WeightInfo::execute(
			*length_bound, // B
			T::MaxMembers::get(), // M
		).saturating_add(proposal.get_dispatch_info().weight), // P
		DispatchClass::Operational
	))]

File: frame/collective/src/lib.rs [hunk 2]
```diff
@@ -292,36 +195,22 @@ decl_module! {
...
-		///   - 1 storage write (codec `O(1)`) for deleting the old `prime` and setting the new one
-		/// # </weight>
-		#[weight = (
+		#[pallet::weight(
 			T::WeightInfo::set_members(
 				*old_count, // M
 				new_members.len() as u32, // N
 				T::MaxProposals::get() // P
-			),
-			DispatchClass::Operational
```
Comment (thiolliere) [inline frame/collective/src/lib.rs:314]: same here, we just need to put the same tuple inside the syntax: pallet::weight()

File: frame/collective/src/lib.rs [hunk 3]
```diff
@@ -683,29 +521,197 @@ decl_module! {
...
+		/// \[proposal_hash, result\]
+		MemberExecuted(T::Hash, DispatchResult),
+		/// A proposal was closed because its threshold was reached or after its duration was up.
+		/// \[proposal_hash, yes, no\]
+		Closed(T::Hash, MemberCount, MemberCount),
+	}
+
+	/// Old name generated by `decl_event`.
+	// TODO :: Have to re-check is "Note message" required ?
+	// #[deprecated(note = "use `Event` instead")]
```
Comment (thiolliere) [inline frame/collective/src/lib.rs:575]: if we keep it I think we can deprecate it
```suggestion
	#[deprecated(note = "use `Event` instead")]
```

File: frame/collective/src/lib.rs (hunk 2 above)
Comment (shamb0) [reply frame/collective/src/lib.rs:314]: Correction Done :+1:

File: frame/collective/src/lib.rs (hunk 1 above)
Comment (shamb0) [reply frame/collective/src/lib.rs:247]: Correction Done :+1:

**Comment by thiolliere** [comment]
so if master merged and a runtime migration either in the pallet or in a polkadot companion at the runtime level and it should be good

**Comment by shamb0** [comment]
> so if master merged and a runtime migration either in the pallet or in a polkadot companion at the runtime level and it should be good

Thanks @thiolliere , will pick up for this week

**Comment by gnunicorn** [comment]
anyone working on this?

**Comment by thiolliere** [comment]
it needs a migration code in polkadot or here. Because the old pallet prefix is different from the pallet name used in polkadot and kusama.

**Comment by shamb0** [comment]
Hello @thiolliere, will pick up run time migration part immediately.

File: frame/collective/src/lib.rs [hunk 4]
```diff
@@ -650,6 +668,12 @@ pub mod pallet {
 	#[pallet::getter(fn prime)]
 	pub type Prime<T: Config<I>, I: 'static = ()> = StorageValue<_, T::AccountId, OptionQuery>;
 
+	/// True if we have upgraded so that AccountInfo contains three types of `RefCount`. False
+	/// (default) if not.
+	#[pallet::storage]
+	pub(super) type UpgradedToTripleRefCount<T: Config<I>, I: 'static = ()> =
```
Comment (thiolliere) [inline frame/collective/src/lib.rs:674]: the name of the storage is not meaningful.
there is no triplerefcount upgrade in this pallet.
`UpgradedPalletPrefix` is more sensible to me

File: frame/collective/src/lib.rs [hunk 5]
```diff
@@ -669,11 +693,30 @@ pub mod pallet {
 	#[pallet::genesis_build]
 	impl<T: Config<I>, I: 'static> GenesisBuild<T, I> for GenesisConfig<T, I> {
 		fn build(&self) {
+			<UpgradedToTripleRefCount<T, I>>::put(false);
```
Comment (thiolliere) [inline frame/collective/src/lib.rs:696]: I would expect to put `true`, if the genesis is build it means there wasn't any storage before, thus we don't need any migration.

File: frame/collective/src/lib.rs [hunk 6]
```diff
@@ -669,11 +693,30 @@ pub mod pallet {
...
 			Pallet::<T, I>::initialize_members(&self.members);
 		}
 	}
 }
 
+mod migrations {
+	use super::*;
+
+	pub fn migrate_to_triple_ref_count<T: Config<I>, I: 'static>() -> frame_support::weights::Weight {
```
Comment (thiolliere) [inline frame/collective/src/lib.rs:705]: name should be more meaningful, maybe `migrate_to_new_pallet_prefix`

File: frame/collective/src/lib.rs [hunk 7]
```diff
@@ -669,11 +693,30 @@ pub mod pallet {
...
+
+		let new_name = T::PalletInfo::name::<Pallet<T, I>>()
+			.expect("Fatal Error Invalid PalletInfo name");
```
Comment (thiolliere) [inline frame/collective/src/lib.rs:708]: we can do:
```rust
if new_name == b"Collective" {
    return 0
}
```
in case the name hasn't changed.

**Review by thiolliere — APPROVED** [review]
migration looks good to me.

A test would be good, something like:
* apply migration
* ensure no keys are remaining at the old prefix, so we haven't miss any storages.

File: frame/collective/src/lib.rs [hunk 8]
```diff
@@ -121,165 +129,78 @@ impl DefaultVote for MoreThanMajorityThenPrimeDefaultVote {
...
 
-// Note that councillor operations are assigned to the operational class.
-decl_module! {
-	pub struct Module<T: Config<I>, I: Instance=DefaultInstance> for enum Call where origin: <T as frame_system::Config>::Origin {
-		type Error = Error<T, I>;
+		fn integrity_test() {
+			T::BlockWeights::get()
+				.validate()
+				.expect("The weights are invalid.");
+		}
```
Comment (thiolliere) [inline frame/collective/src/lib.rs:199]: this is new ? why do we do this now ?

File: frame/collective/src/lib.rs (hunk 4 above)
Comment (shamb0) [reply frame/collective/src/lib.rs:674]: Sounds good, Sure will modify to `UpgradedPalletPrefix`. 
I assumed `UpgradedToTripleRefCount` is coined after runtime version 
info like `Major.Minor.Subversion`.

File: frame/collective/src/lib.rs (hunk 5 above)
Comment (shamb0) [reply frame/collective/src/lib.rs:696]: Make sense, Done :+1:

File: frame/collective/src/lib.rs (hunk 6 above)
Comment (shamb0) [reply frame/collective/src/lib.rs:705]: Make sense, Done :+1:

File: frame/collective/src/lib.rs (hunk 7 above)
Comment (shamb0) [reply frame/collective/src/lib.rs:708]: Indeed make sense Done :+1:

File: frame/collective/src/lib.rs (hunk 8 above)
Comment (shamb0) [reply frame/collective/src/lib.rs:199]: Sure will remove it, Sorry its my bad, took it from `frame/system/src/lib.rs`,
assumed that it's part of trait Hooks<> pattern.

File: frame/system/src/lib.rs [hunk 9]
```diff
@@ -269,12 +269,6 @@ pub mod pallet {
 				0
 			}
 		}
-
-		fn integrity_test() {
-			T::BlockWeights::get()
-				.validate()
-				.expect("The weights are invalid.");
-		}
```
Comment (thiolliere) [inline frame/system/src/lib.rs:277]: why do you remove it ?

In general I think ppl should review their PR and asking themself the question:
* is this change part of the PR ?
* is it just a not needed extra (in this case you can put a comment saying that it is an extra not really part of the PR, and the reason why it is done)

**Comment by thiolliere** [comment]
can you merge master ? Once this is done I'll take a second look and it needs another approval maybe @ascjones or somebody else. Then it is good to go.

**Comment by shamb0** [comment]
Hello @thiolliere, merge to upstream master Done :+1:

File: frame/collective/src/lib.rs [hunk 10]
```diff
@@ -694,29 +551,236 @@ decl_module! {
...
+
+	pub fn migrate_to_new_pallet_prefix<T: Config<I>, I: 'static>() -> frame_support::weights::Weight {
+
+		let new_name = T::PalletInfo::name::<Pallet<T, I>>()
+			.expect("Fatal Error Invalid PalletInfo name")
+			.as_bytes();
+
+		if new_name == b"Collective" {
+			return 0
+		}
```
Comment (thiolliere) [inline frame/collective/src/lib.rs:715]: ah but this is wrong, the old prefix is `Collective` only for the default instance of the pallet.

For the instance 1 the prefix is `Instance1Collective` or something like this.
There is no way to know the old pallet prefix as `I` no longer bounds `Instance` trait.
```suggestion
```

File: frame/collective/src/lib.rs [hunk 11]
```diff
@@ -694,29 +551,236 @@ decl_module! {
...
+		T::BlockWeights::get().max_block
+	}
+}
+
+#[cfg(feature = "std")]
+impl<T: Config<I>, I: 'static> GenesisConfig<T, I> {
+	/// Direct implementation of `GenesisBuild::build_storage`.
+	///
+	/// Kept in order not to break dependency.
+	pub fn build_storage(&self) -> Result<sp_runtime::Storage, String> {
```
Comment (thiolliere) [inline frame/collective/src/lib.rs:733]: actually we can deprecate them
```suggestion
	#[deprecated(note = "use [`GenesisBuild::build_storage`] instead")]
	pub fn build_storage(&self) -> Result<sp_runtime::Storage, String> {
```

File: frame/collective/src/lib.rs [hunk 12]
```diff
@@ -694,29 +551,236 @@ decl_module! {
...
+		<Self as GenesisBuild<T, I>>::build_storage(self)
+	}
+
+	/// Direct implementation of `GenesisBuild::assimilate_storage`.
+	///
+	/// Kept in order not to break dependency.
+	pub fn assimilate_storage(
```
Comment (thiolliere) [inline frame/collective/src/lib.rs:740]:
```suggestion
	#[deprecated(note = "use [`GenesisBuild::assimilate_storage`] instead")]
	pub fn assimilate_storage(
```

**Review by thiolliere — APPROVED** [review]

**Comment by shawntabrizi** [comment]
@shamb0 looks like collective got updated since this was done.

Care to update this for latest master?

If so, feel free to reopen

**Comment by thiolliere** [comment]
yes need master merge, and a second review
